### PR TITLE
Update NNPI Backend to v0.4.0.1

### PR DESCRIPTION
**Pull request by omromano**
Summary: Updating the NNPI Backend to v0.4.0.1

### CODE REVIEW

**Comment by jsubag** [comment]
@jfix71 , @arunm-git Please take a look.

File: lib/Backends/NNPI/NNPIMLTraceWrapper.cpp [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+#include <sstream>
+#include <time.h>
+#include <unistd.h>
+#include <unordered_map>
+#include <vector>
+
+#define MAX_TRACE_BUFFER_SIZE (1024 * 1024 * 5)
+#define TRACE_READ_BUFFER_SIZE (1024 * 10)
+
+static inline uint64_t secondsToMicrosecons(double seconds) {
```
Comment (jfix71) [inline lib/Backends/NNPI/NNPIMLTraceWrapper.cpp]: nit: `Microseconds` (missing d)

File: lib/Backends/NNPI/NNPIMLTraceWrapper.cpp [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+  return (uint64_t)(seconds * 1e6f);
+}
+
+static inline int64_t nanosecondsToMicroseconsSigned(int64_t nanoseconds) {
```
Comment (jfix71) [inline lib/Backends/NNPI/NNPIMLTraceWrapper.cpp]: nit: `Microseconds` (missing d)

File: lib/Backends/NNPI/NNPIMLTraceWrapper.cpp [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+  return nanoseconds / 1e3;
+}
+
+static uint64_t getUpRefTime() {
+  static double refTime = 0;
```
Comment (jfix71) [inline lib/Backends/NNPI/NNPIMLTraceWrapper.cpp]: Bit confused by this -- looks like it'll never be updated?

File: lib/Backends/NNPI/NNPIMLTraceWrapper.cpp [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+  if (refTime > 0) {
+    return refTime;
+  }
+  // Get uptime.
+  double uptime(0.0);
+  if (std::ifstream("/proc/uptime", std::ios::in) >> uptime) {
+    return secondsToMicrosecons(uptime);
+  }
+  return 0.0;
```
Comment (jfix71) [inline lib/Backends/NNPI/NNPIMLTraceWrapper.cpp]: tiny nit: `return 0` as this is `uint64_t`.

File: lib/Backends/NNPI/NNPIMLTraceWrapper.cpp [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+
+  void parseLine(std::string line, NNPITraceEntry &entry) {
+    size_t idx = 0;
+    std::istringstream linestream(line);
+    do {
+      std::string part;
+      linestream >> part;
+
+      switch (idx) {
+      case 0: // Pid.
```
Comment (jfix71) [inline lib/Backends/NNPI/NNPIMLTraceWrapper.cpp]: Can we create an enum for these to make it a little cleaner?

File: lib/Backends/NNPI/NNPIMLTraceWrapper.cpp [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+      return false;
+    }
+  }
+
+  // Handle stream.
+  std::string line;
+  NNPITraceParser parser(0, upRefTime_);
+  int64_t timeDiffSec = parser.getTimeDiffSec();
+  while (std::getline(inputStream, line)) {
+    if (line.find("#", 0) == 0)
```
Comment (jfix71) [inline lib/Backends/NNPI/NNPIMLTraceWrapper.cpp]: nit: We always use curly braces even on single line if/while/etc.

File: lib/Backends/NNPI/NNPIMLTraceWrapper.cpp [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+        break;
+      }
+      case 4: // FUNCTION.
+      {
+        entry.traceType = getType(part);
+        break;
+      }
+      case 5: // May also be function for NNPI_TRACE_MARK.
+      {
+        if (entry.traceType == NNPI_TRACE_MARK &&
```
Comment (jfix71) [inline lib/Backends/NNPI/NNPIMLTraceWrapper.cpp]: Little confused by this -- isn't `entry` always going to be default constructed here to `NNPI_TRACE_UNKNOWN`, at least by current use cases (i.e. called from `NNPITraceContext::load()`)? Perhaps will be used differently in the future?

File: lib/Backends/NNPI/NNPIMLTraceWrapper.cpp [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+}
+
+bool NNPITraceContext::load() {
+  entries_.clear();
+  std::stringstream inputStream;
+  uint32_t size = TRACE_READ_BUFFER_SIZE;
+  uint32_t actualSize = size;
+  char readData[TRACE_READ_BUFFER_SIZE + 1];
+
+  // Read trace bytes into steamd.
```
Comment (jfix71) [inline lib/Backends/NNPI/NNPIMLTraceWrapper.cpp]: typo: `steamd`

File: lib/Backends/NNPI/NNPICompiledFunction.cpp [hunk 9]
```diff
@@ -65,6 +65,11 @@ Error NNPICompiledFunction::compile(Function *F, const BackendOptions &opts) {
     }
   }
 
+  const auto numCores = EnvIceCores();
+  if (!numCores.empty()) {
+    config_.numCoresToUse = std::stoul(numCores);
```
Comment (jsubag) [inline lib/Backends/NNPI/NNPICompiledFunction.cpp]: @nrsatish This is the place to set the "numCoresToUse" for the compiler

File: lib/Backends/NNPI/NNPIMLTraceWrapper.cpp (hunk 7 above)
Comment (omromano) [reply lib/Backends/NNPI/NNPIMLTraceWrapper.cpp]: @jfix71 Point taken. Added a comment to make it a little clearer and will revisit the parsing in next versions.

**Comment by omromano** [comment]
@jfix71 Made changes based on your comments.

**Pull request closed without merging**